packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -963,6 +963,16 @@ class _DatePickerHeader extends StatelessWidget {
   }
 }
 
+/// Signature for predicating dates for enabled date selections in range pickers.
+///
+/// The [selectedStartDay] and [selectedEndDay] are the currently selected start and end dates of a date range, which permits
+/// to enable or disable dates based on the user selection. (Example: in a hostel's room selection,
+/// you are not able to select the end date after the next non-selectable day).
+///
+/// See [showDateRangePicker], which has a [SelectableDayForRangePredicate] parameter used
+/// to specify allowable days in the date range picker.
+typedef SelectableDayForRangePredicate = bool Function(DateTime day, DateTime? selectedStartDay, DateTime? selectedEndDay);
```

Comment:
ping @cedvdb 

it can also prevent the user of selecting the same end date as the start date, by making the selected start day as non-selectable.

here is my implementation as an example to make the provided videos working (without the `inclusiveEndDate`, which I'm ok with you, that seems use-case specific!):
```dart
Future<DateTimeRange?> showDateRangeDialog(
  BuildContext context, {
  DateTimeRange? initialSelectedRange,
  Iterable<DateTimeRange> unavailableRanges = const [],
}) async {
  final today = DateTime.now();
  final availabilityChecker = DayAvailabilityChecker(unavailableRanges);

  return showDateRangePicker(
    context: context,
    firstDate: today,
    lastDate: today.add(const Duration(days: 60)),
    initialDateRange: initialSelectedRange,
    selectableDayPredicate: (day, selectedStartDate, selectedEndDate) {
      if (selectedEndDate == null && selectedStartDate != null && day.isBefore(selectedStartDate)) {
        final nextNonAvailableDay = availabilityChecker.nextNonAvailableDay(selectedStartDate)?.add(const Duration(days: 1));
        return nextNonAvailableDay == null || day.isBefore(nextNonAvailableDay);
      }
      return availabilityChecker.isAvailable(day);
    },
  );
}
```

Reply:
Yes, this is much better imo, this way all use cases are covered.

Reply:
Nice. How to get another reviews to agree on the api (if needed) ?

Now I'll make the tests, but I never did frontend tests, even less flutter tests (I'm a backend engineer 😅). Which new tests should I make at the end in your opinion ?

Reply:
I'd write 2 tests:

- 1 test that verifies that when the predicate is false, the date is not selectable
- 1 test that verifies that when the predicate is updated (after the start selection has been made for example) that the new predicate result is reflected

Reply:
I added some tests.

missing tests:
- input mode validation
- dynamically changed selectableDayPredicate returns are reflected to the ui when selecting days (i don't know how to check if a date is not clickable or greyed)

I'm not able of expecting an assertion error if you can help me (checking the given first/last/initial dates that are all selectable) 🙏 The error is thrown, but seems to not be catched by the `expect` 🤔

Reply:
> i don't know how to check if a date is not clickable or greyed

if it's unselectable, pressing a day will do nothing. so you can test that the selection did not change after tapping that day.

Reply:
Yess it's what I'm doing here :
```dart
expect(() async {
      await preparePicker(tester, (Future<DateTimeRange?> range) async {
        await range;
      }, selectableDayPredicate: (DateTime day, _, __) => day.day != 12);
    }, throwsAssertionError);
```

the doc states about the following, that should be okay 🤔 
```
/// This is equivalent to `throwsA(isA<AssertionError>())`.
/// ...
final Matcher throwsAssertionError = throwsA(isAssertionError);

Reply:
also not working with `expectLater`

Reply:
Why are you testing that it throws an assertion error ?

Reply:
Here is the complete test, it may be more clear:
```dart
  testWidgets('Cannot show date range picker if firstDate is not selectable', (WidgetTester tester) async {
    initialDateRange = null;
    firstDate = DateTime(2017, DateTime.january, 12);
    lastDate = DateTime(2017, DateTime.january, 13);

    expect(() async {
      await preparePicker(tester, (Future<DateTimeRange?> range) async {
        await range;
      }, selectableDayPredicate: (DateTime day, _, __) => day.day != 12);
    }, throwsAssertionError);
  });
```

Reply:
why are you expecting an assertion error ? An assertion error happens when  there is something like `assert(3 == 2)`.

Reply:
You might need to execute the function, inside, I always forget which one it is, try
```
final nameThis = () async {
      await preparePicker(tester, (Future<DateTimeRange?> range) async {
        await range;
      }, selectableDayPredicate: (DateTime day, _, __) => day.day != 12);
    };
expect(nameThis(), throwsAssertionError);
```

Reply:
Ok I removed them, nothing is working as excepted, the test tooling is not expecting an exception. Also, there is no assertion tests in the date or date_range pickers, so I assume it's ok.

I removed the assertion of first and last dates, as we may want to display a picker starting from a date and ending with dates that could not be selectable (if the hotel's room is not available by example).


But I still don't succeed to make the test work for testing a changing predicate (it's working in my project, but totally not in the tests). The idrea is to make selectable all the dates by default, but when you start picking a range (start date not null, but end date null), then the dates before the start date are not selectable.

Here is what I tried without success:
```dart
testWidgets('Selectable date becoming non selectable when selected start day', (WidgetTester tester) async {
    await preparePicker(tester, (Future<DateTimeRange?> range) async {
      await tester.tap(find.text('12').first);
      await tester.tap(find.text('11').first);
      await tester.tap(find.text('14').first);
      await tester.tap(find.text('SAVE'));
      expect(await range, DateTimeRange(
        start: DateTime(2016, DateTime.january, 12),
        end: DateTime(2016, DateTime.january, 14),
      ));
    }, selectableDayPredicate: (DateTime day, DateTime? selectedStart, DateTime? selectedEnd) {
      if (selectedEnd == null && selectedStart != null) {
        return day == selectedStart || day.isAfter(selectedStart);
      }
      return true;
    });
  });
```

Reply:
> I removed the assertion of first and last dates, as we may want to display a picker starting from a date and ending with dates that could not be selectable (if the hotel's room is not available by example).
 
 I guess that makes sense yeah
 
 > without success
 
 What's without success ? error ? unexpected result ?
 
 You might be missing a `pumpAndSettle()` ?
 
 ```dart
      await tester.tap(find.text('12').first);
      await tester.pumpAndSettlle();
      await tester.tap(find.text('11').first);
      await tester.pumpAndSettlle();
      await tester.tap(find.text('14').first);
      await tester.pumpAndSettlle();
      await tester.tap(find.text('SAVE'));
      await tester.pumpAndSettlle();
```

Reply:
But why the other tests doesn't need it? It stops working when tapping on the day 11

Reply:
This is the only test failing (or taking forever, it depends). I'll try this pumpAndSettle

Reply:
And let's make the customization PR when this PR is merged! 😁

Reply:
It's working with your `pumpAndSettle`, so I suppose that the tap are done too quick to let the ui being redraw.

Putting the PR as ready now!